dali/fuzzing/dali_harness.h

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_FUZZING_DALI_HARNESS_H_
+#define DALI_FUZZING_DALI_HARNESS_H_
+
+#include <string>
+#include <vector>
+#include <utility>
+
+#include "dali/core/common.h"
+#include "dali/pipeline/pipeline.h"
+#include "dali/util/image.h"
+#include "dali/test/dali_test_config.h"
+
+namespace dali {
+class DecoderHarness {
+ public:
+  DecoderHarness(
+    string &path,
+    string file_extension = ".jpg",
+    int batch_size = 4,
+    int num_threads = 4,
+    int device_id = 0) :
+      batch_size_(batch_size),
+      pipeline_(batch_size, num_threads, device_id) {
+    const string jpeg_folder = make_string(
+      testing::dali_extra_path(),
+      "/db/fuzzing/");
+    image_names_ = ImageList(jpeg_folder, {file_extension}, batch_size_ - 1);
+    image_names_.push_back(path);
+    LoadImages(image_names_, &images_);
+
+    MakeBatch();
+    SetupPipeline();
+  }
+
+  void MakeBatch() {
+    TensorListShape<> shape(batch_size_, 1);
+    for (int i = 0; i < batch_size_; ++i) {
+      shape.set_tensor_shape(i, {images_.sizes_[i]});
+    }
+
+    input_data_.template mutable_data<uint8>();
+    input_data_.Resize(shape);
+
+    for (int i = 0; i < batch_size_; ++i) {
+      std::memcpy(
+        input_data_.template mutable_tensor<uint8>(i),
+        images_.data_[i],
+        images_.sizes_[i]);
+      input_data_.SetSourceInfo(i, image_names_[i] + "_" + std::to_string(i));
+    }
+  }
+
+  virtual void SetupPipeline() {
+    pipeline_.AddExternalInput("raw_images");
+    pipeline_.SetExternalInput("raw_images", input_data_);
+
+    pipeline_.AddOperator(
+      OpSpec("ImageDecoder")
+        .AddArg("device", "mixed")
+        .AddArg("output_type", DALI_RGB)
+        .AddInput("raw_images", "cpu")
+        .AddOutput("images", "gpu"));
+    pipeline_.Build({{"images", "gpu"}});
+  }
+
+  void Run() {
+    DeviceWorkspace ws;
+    pipeline_.RunCPU();
+    pipeline_.RunGPU();
+    pipeline_.Outputs(&ws);
+  }
+
+ protected:
+  int batch_size_;
+  Pipeline pipeline_;
+  TensorList<CPUBackend> input_data_;
+
+  vector<string> image_names_;
+  ImgSetDescr images_;
+};
+
+class ResNetHarness : public DecoderHarness {
```

Comment:
I'm not sure if it should inherit from `DecoderHarness`. Maybe it should be a basic harness instead and `DecoderHarness` would inherit from it as well.

Reply:
I extracted common functionality to `FileListHarness` - base for harnesses that want to read from file_list and pass the path to additional file. Derived harnesses only change how pipeline is defined and what file extensions they want to read.
In the future we may make `FileListHarness` inherit from `BaseHarness` when we have other harnesses getting data some other way. Is that ok?